node/gum/src/lib.rs

```diff
@@ -112,7 +113,55 @@ pub use jaeger::hash_to_trace_identifier;
 #[doc(hidden)]
 pub use polkadot_primitives::{CandidateHash, Hash};
 
-pub use gum_proc_macro::{debug, error, info, trace, warn};
+pub use gum_proc_macro::{debug, error, info, trace, warn, warn_if_frequent};
 
 #[cfg(test)]
 mod tests;
+
+/// Max allowed frequency of debug messages
+pub const MAX_FREQ_RATE: f64 = 1.0;
+const MAX_FREQ_SIZE: usize = 10;
+
+/// Utility struct to compare the rate of its own calls.
+pub struct Freq {
+	timestamps: VecDeque<u64>,
+}
+
+impl Freq {
+	/// Initiates a new instanse
+	pub fn new() -> Self {
+		Self { timestamps: VecDeque::with_capacity(MAX_FREQ_SIZE) }
+	}
+
+	/// Compares the rate of its own calls with the passed one.
+	pub fn is_frequent(&mut self, max_rate: f64) -> bool {
+		self.record();
+
+		// Two attempts is not enough to call something as frequent.
+		if self.timestamps.len() < 3 {
+			return false
+		}
+
+		let elapsed =
+			self.timestamps.back().expect("Checked above, timestamps include >=3 entries") -
+				self.timestamps.front().expect("Checked above, timestamps include >=3 entries");
+		if elapsed == 0 {
+			return true // More then 1000 times per second is frequent enough
+		}
+
+		let rate = (self.timestamps.len() as u64 * 1000 / elapsed) as f64;
+		rate > max_rate
+	}
+
+	fn record(&mut self) {
+		while self.timestamps.len() >= MAX_FREQ_SIZE {
+			let _ = self.timestamps.pop_front();
+		}
+
+		let now = std::time::SystemTime::now()
```

Comment:
We should use coarse timers here, `SystemTime::now()` call is expensive and we don't want it in any hot path (e.g. in the logging). Maybe we should have a separate implementation for the unit tests in this case.